Shopify/ruby-lsp
# Migrate chruby activation to the new version manager implementation

### Motivation

The changes in https://github.com/Shopify/vscode-ruby-lsp/pull/923 were useful for testing purposes, but I think we cannot have a blanket implementation that will work for every version manager. Clearly, they each have their own specificities that need to be accounted for with care (e.g.: Mise supports multiple level of hierarchy in configuration and RVM supports the .ruby-gemset file).

To move forward with more confidence, I'm proposing a similar approach, with a redesign, so that we can ship this as part of our stable releases.

### Implementation

The idea of the implementation is that

- We keep the version manager setting, so that people can specify what they are using
- We make each version manager a separate class, which is responsible for processing the configuration files and finding the Ruby installation directory (which are both different in most managers)
- The parent version manager enforces what needs to be implemented by each version manager and puts all the pieces together to activate the Ruby environment

This way, we can also move slowly and ship one version manager at a time, gathering feedback as we go.

This maintains the similar advantages of the other PR, such as fully decoupling from shells and allowing us to have extensive testing for each manager. Additionally, this PR implements most of the shared behaviour, so the next managers should be easier to implement.

### Automated Tests

Added tests for chruby.

### Manual Tests

1. Launch the extension in a project using chruby
2. Verify that the LSP launches correctly
